reactive/spark.py

```diff
@@ -32,59 +35,113 @@ def report_blocked():
 
 @when('hadoop.related')
 @when_not('hadoop.ready')
-def report_waiting(hadoop):
+def report_waiting(hadoop):# pylint: disable=w0613
     hookenv.status_set('waiting', 'Waiting for Hadoop to become ready')
 
 
 # TODO: support standalone mode when Yarn not connected
 @when('hadoop.ready')
 @when_not('spark.installed')
-def install_spark(hadoop):
-
-    dist = get_dist_config()
+def install_spark(hadoop):# pylint: disable=w0613
+    dist = DistConfig(data=layer.options('apache-spark'))
     spark = Spark(dist)
     if spark.verify_resources():
         hookenv.status_set('maintenance', 'Installing Apache Spark')
-        dist.add_dirs()
-        dist.add_packages()
         spark.install()
         spark.setup_spark_config()
         spark.install_demo()
         set_state('spark.installed')
 
 
-@when('spark.installed', 'hadoop.ready')
+@when('hadoop.ready', 'spark.installed')
+@when_not('livy.installed')
+def install_livy(hadoop):# pylint: disable=w0613
+    dist = DistConfig(data=layer.options('livy'))
+    livy = Livy(dist)
+    if livy.verify_resources():
+        hookenv.status_set('maintenance', 'Installing Livy REST server')
+        livy.install()
+        set_state('livy.installed')
+
+
+@when('spark.installed', 'livy.installed', 'hadoop.ready')
 @when_not('spark.started')
-def start_spark(hadoop):
+def start_spark(hadoop):# pylint: disable=w0613
     hookenv.status_set('maintenance', 'Setting up Apache Spark')
-    spark = Spark(get_dist_config())
-
+    dist = DistConfig(data=layer.options('apache-spark'))
+    spark = Spark(dist)
     spark.configure()
     spark.start()
     spark.open_ports()
     set_state('spark.started')
+
+
+@when('spark.installed', 'livy.installed', 'hadoop.ready', 'spark.started')
+@when_not('livy.started')
+def start_livy(hadoop):# pylint: disable=w0613
+    hookenv.status_set('maintenance', 'Setting up Livy REST server')
+    dist = DistConfig(data=layer.options('livy'))
+    livy = Livy(dist)
+    mode = hookenv.config()['spark_execution_mode']
+    livy.configure(mode)
+    livy.start()
+    livy.open_ports()
+    set_state('livy.started')
+    hookenv.status_set('active', 'Ready')
+
+
+@when('spark.installed', 'livy.installed', 'hadoop.ready', 'spark.started', 'livy.started')
+def reconfigure_spark(hadoop):# pylint: disable=w0613
+    config = hookenv.config()
+    if not data_changed('configuration', config):
+        return
```

Comment:
I'd like to suggest adding `@when('config.changed')` instead if we're not watching a specific subset of config.